Add E2E tests

xmudrii opened the pull request:
This PR adds E2E tests for the EtcdProxyController.

The E2E tests assumes you have a working Kubernetes cluster and EtcdProxy Controller deployed. The `hack/run-e2e.sh` script, which is used by CI, deploys the controller from the deployment manifest located in the `artifacts` directory.

By default, tests use the kubeconfig located in `$HOME/.kube/config`. The kubeconfig path can be changed by modifying the `kubeconfigPath` variable in the `test/e2e/e2e_test.go` file.

The tests are located in the `test/e2e/e2e_test.go` file. Currently, we have one test function which tests:
* Is EtcdStorage creation successful,
* Are EtcdStorage conditions set as intended,
* Is ReplicaSet with the expected name deployed,
* Is expected number of pods available (running),
* Is Service with the expected name created.

We're testing two cases for now:
* EtcdStorage is valid,
* EtcdStorage name is too long.

I'm also working on testing are all resources going to be deleted once the EtcdStorage object is deleted.

xmudrii commented on Makefile:
```diff
@@ -75,13 +76,17 @@ build-darwin-amd64: ## Create the etcdproxy-controller executable for Darwin (os
-verify-ci: install-tools ## Run code checks
-	PKGS="${GOFILES}" GOFMT="gofmt" ./hack/verify-ci.sh
+.PHONY: test-e2e
+test-e2e: ## Run E2E tests. Requires working Kubernetes cluster and a Kubeconfig file.
```
This should be updated to use the `run-e2e.sh` script.

A follow-up question: do we need to leave a note that this is potentially destructive operation and to be sure you're running it on the correct cluster?

deads2k replied:
> A follow-up question: do we need to leave a note that this is potentially destructive operation and to be sure you're running it on the correct cluster?

note that your tests may be destructive.  We can assume these people know what they're doing and get what they deserve.

xmudrii replied:
Added a note and fixed the command.

xmudrii commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+)
+
+// kubeconfigPath is path to the kubernetes configuration file.
+var kubeconfigPath = filepath.Join(os.Getenv("HOME"), "/.kube/config")
```
Is there a better way for providing `kubeconfigPath`? Maybe with an environment variable, or this is good enough?

deads2k replied:
> Is there a better way for providing kubeconfigPath? Maybe with an environment variable, or this is good enough?

We respect `KUBECONFIG` in most (all?) kube binaries.  I'd use that.

xmudrii replied:
This is a great idea, updated the PR to use it.

xmudrii commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+var kubeconfigPath = filepath.Join(os.Getenv("HOME"), "/.kube/config")
+
+// newKubeClient create new Kubernetes client instance based on kubeconfig file.
+func newKubeClient(kubeconfigPath string) (*kubernetes.Clientset, error) {
```
Is it okay to leave those functions in this file, or should I create a new file such as `util.go`?

deads2k replied:
> Is it okay to leave those functions in this file, or should I create a new file such as util.go?

don't create a "util" anything until absolutely necessary.  Let's say 1000 lines.  They generally become a bucket of stuff.

xmudrii commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+	return clientset.NewForConfig(config)
+}
+
+func replicasetName(etcdstorageName string) string {
```
My idea was to have a helper functions for calculating names, instead of hardcoding them in test cases, to make potential changes easier.

deads2k replied:
> My idea was to have a helper functions for calculating names, instead of hardcoding them in test cases, to make potential changes easier.

I generally question the utility of one liners.  Do you think it's likely we'll change our minds?  How many places would we really have to update it?  *Should* such a change be hard since we're changing external behavior?

xmudrii replied:
> Do you think it's likely we'll change our minds? 

I don't think we'll ever change those. Even if we do in some case, that would be just a few lines to update. Let's get rid of those functions.

xmudrii commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+	return fmt.Sprintf("etcd-%s", etcdstorageName)
+}
+
+func TestDeployEtcdStorage(t *testing.T) {
```
Is it okay to have all these checks in this function, or I should split them to multiple tests? In this function we're testing is the EtcdStorage instance deployed, and are all relevant resources deployed.

TODO: add godoc comment

xmudrii replied:
Added godoc comment.

xmudrii commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+		}
+
+		// Only continue if etcdStorage is expected to be valid.
+		if tc.etcdStorageValid {
```
I have a problem here.

We have defined that EtcdStorages name length should be less than 59 characters (64 which is max for Services, -5 for `etcd-` prefix). I added a test case for that, but, this is going to fail on `Create`. 

But that failure is expected, so I was not sure how to stop there. I added `etcdStorageValid` bool, and if it is false, error on Create is going to be ignored, and we're not going to continue.

Is there a better way to mark test as passed, than this one? There is 'FailNow()`, which fails, but I don't want fail`.
There is also `SkipNow()`, but it marks test as skipped, and I was not sure do we want that.

xmudrii commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+		// Only continue if etcdStorage is expected to be valid.
+		if tc.etcdStorageValid {
+			// It takes some time for the ReplicaSet and Service to be created, so we need to poll.
+			err = wait.Poll(500*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
```
TODO: check for error.

xmudrii replied:
Fixed.

xmudrii commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+				t.Fatalf("expected replicaset '%s', but got: %v", tc.expectedReplicaSetName, err)
+			}
+			// Wait for pods to become ready.
+			err = wait.Poll(500*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
```
TODO: check for error.

xmudrii replied:
Fixed.

xmudrii commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+			}
+
+			// Check is ReplicaSet deployed.
+			rs, err := client.AppsV1().ReplicaSets("kube-apiserver-storage").Get(tc.expectedReplicaSetName, metav1.GetOptions{})
```
This is not needed, we have it in `wait.Poll`.

xmudrii replied:
Fixed.

deads2k commented on .travis.yml:
```diff
@@ -3,14 +3,38 @@ language: go
+      script: make test-e2e
+      before_script:
+        # Download kubectl, which is a requirement for using minikube and running e2e tests.
+        - curl -Lo kubectl https://storage.googleapis.com/kubernetes-release/release/${KUBERNETES_VERSION}/bin/linux/amd64/kubectl && chmod +x kubectl && sudo mv kubectl /usr/local/bin/
```
@soltysh @liggitt next release I'm solving my oc problems like this! :)

soltysh replied:
I object :stuck_out_tongue_closed_eyes:

deads2k commented on .travis.yml:
```diff
@@ -3,14 +3,38 @@ language: go
+        # Update the kubeconfig to use the minikube cluster.
+        - minikube update-context
+        # Wait for Kubernetes to be up and ready.
+        - JSONPATH='{range .items[*]}{@.metadata.name}:{range @.status.conditions[*]}{@.type}={@.status};{end}{end}'; until kubectl get nodes -o jsonpath="$JSONPATH" 2>&1 | grep -q "Ready=True"; do sleep 1; done
```
In addition, I'd suggest waiting for the server's /healthz to report true.

xmudrii replied:
Added a TODO note. I'll try to get it for this PR.

deads2k commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,147 @@
+	return kubernetes.NewForConfig(config)
+}
+
+func newEtcdProxyClinet(kubeconfigPath string) (*clientset.Clientset, error) {
```
typo

xmudrii replied:
:see_no_evil:

deads2k commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,137 @@
+
+// newEtcdProxyClient creates new client for etcdstorage resources.
+func newEtcdProxyClient() (*clientset.Clientset, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", os.Getenv("KUBECONFIG"))
```
I would have one helper to create the config and then do the NewForConfig call inline.  I also tolerate `NewForConfigOrDie` in tests.

deads2k commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,137 @@
+			etcdStorageValid:       true,
+		},
+		{
+			name: "test etcdstorage creation - name too long",
```
this should be a unit test

deads2k commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,137 @@
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			// Create an EtcdStorage instance.
+			es, err := etcdproxyClient.EtcdV1alpha1().EtcdStorages().Create(tc.etcdStorage)
```
missing deletion/cleanup

xmudrii replied:
Added.

deads2k commented:
This is fine, but in general I expect my e2e tests to tell a story.  In kube, they alias `kubectl` and use an exec helper to instantiate manifests and check status on that.

The ideal e2e test would be creating your sample manifests.  If you want to simply create a bash script you call to do that for now, that would be fine, but I think it is important to do.

xmudrii commented:
@deads2k I've addressed comments from yesterday and added an E2E story for the sample-apiserver. Compared to EtcdStorage story, which is using a Bash script + Go, the sample-apiserver story is only using the bash script.

The tests for sample-apiserver are checking is the pod ready. This is probably not enough, but I don't have idea what else I could check. If you have an idea, let me know. :)

Also, I've disabled E2E tests in Travis as they're failing. The tests are passing locally, so for now, let's leave it like this if you agree.

deads2k commented on hack/run-e2e.sh:
```diff
@@ -0,0 +1,66 @@
+    READY_REPLICAS=$(kubectl get rs apiserver -o jsonpath="{.status.readyReplicas}")
+done
+
+echo '* Starting sample-apiserver e2e tests.'
```
followup issue to actually create a resource.

deads2k commented on test/e2e/e2e_test.go:
```diff
@@ -0,0 +1,127 @@
+	// Initialize clients.
+	cfg, err := clientcmd.BuildConfigFromFlags("", os.Getenv("KUBECONFIG"))
+	if err != nil {
+		t.Fatalf("unable to create kubernetes config from provided kubeconfig file: %v", err)
```
nit: I generally just use t.Fatal(err).  You get a line number, so you can track it back easily enough.

deads2k commented:
> Also, I've disabled E2E tests in Travis as they're failing. The tests are passing locally, so for now, let's leave it off if you agree.

Yeah, that's fine.  Don't forget to run your e2e and tell us in the description whether it worked or not.
